File: problems/2580_count_ways_to_group_overlapping_ranges/solution.py
from typing import List


class Solution:
    def countWays(self, ranges: List[List[int]]) -> int:
        MOD = 10**9 + 7
        ranges.sort()
        groups = 0
        end = -1
        for s, e in ranges:
            if s > end:
                groups += 1
            end = max(end, e)
        return pow(2, groups, MOD)

File: problems/2580_count_ways_to_group_overlapping_ranges/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int countWays(vector<vector<int>>& ranges) {
        sort(ranges.begin(), ranges.end());
        long long MOD = 1e9+7;
        int groups = 0, end = -1;
        for (auto& r : ranges) { if (r[0] > end) groups++; end = max(end, r[1]); }
        long long ans = 1;
        for (int i = 0; i < groups; i++) ans = ans * 2 % MOD;
        return ans;
    }
};
